File: src/utils/passwords.js
const crypto = require('crypto');

function generatePassword({ length = 16, upper = true, lower = true, digits = true, symbols = true } = {}) {
  const sets = [];
  if (upper) sets.push('ABCDEFGHIJKLMNOPQRSTUVWXYZ');
  if (lower) sets.push('abcdefghijklmnopqrstuvwxyz');
  if (digits) sets.push('0123456789');
  if (symbols) sets.push('!@#$%^&*()-_=+[]{}|;:,.<>?');
  if (sets.length === 0) throw new Error('At least one character set required');
  const all = sets.join('');
  const bytes = crypto.randomBytes(length);
  let out = '';
  for (let i = 0; i < length; i++) {
    out += all[bytes[i] % all.length];
  }
  return out;
}

function entropyBits(password) {
  // estimate entropy by log2(possibleSymbols^length) where possibleSymbols is count of used charsets
  const hasUpper = /[A-Z]/.test(password);
  const hasLower = /[a-z]/.test(password);
  const hasDigits = /[0-9]/.test(password);
  const hasSymbols = /[^A-Za-z0-9]/.test(password);
  let pool = 0;
  if (hasUpper) pool += 26;
  if (hasLower) pool += 26;
  if (hasDigits) pool += 10;
  if (hasSymbols) pool += 32; // approximate symbol count
  if (pool === 0) return 0;
  return Math.log2(Math.pow(pool, password.length));
}

module.exports = { generatePassword, entropyBits };

File: src/utils/passwords.py
"""
Password generation and entropy calculation utilities.
"""

import secrets
import re
import math


def generate_password(length=16, upper=True, lower=True, digits=True, symbols=True):
    """
    Generate a random password with specified character sets.

    Args:
        length: password length (default 16)
        upper: include uppercase letters
        lower: include lowercase letters
        digits: include digits
        symbols: include symbols

    Returns:
        Generated password string
    """
    sets = []
    if upper:
        sets.append('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
    if lower:
        sets.append('abcdefghijklmnopqrstuvwxyz')
    if digits:
        sets.append('0123456789')
    if symbols:
        sets.append('!@#$%^&*()-_=+[]{}|;:,.<>?')

    if not sets:
        raise ValueError('At least one character set required')

    if length < len(sets):
        raise ValueError(f'Length must be at least {len(sets)} to satisfy all character requirements')

    all_chars = ''.join(sets)

    # Guarantee one character from each required set, then fill the rest randomly
    guaranteed = [secrets.choice(s) for s in sets]
    filler = [secrets.choice(all_chars) for _ in range(length - len(guaranteed))]

    password_chars = guaranteed + filler
    secrets.SystemRandom().shuffle(password_chars)

    return ''.join(password_chars)


def entropy_bits(password):
    """
    Estimate entropy of password by calculating log2(possibleSymbols^length).

    Args:
        password: password string to analyze

    Returns:
        Estimated entropy in bits
    """
    has_upper = bool(re.search(r'[A-Z]', password))
    has_lower = bool(re.search(r'[a-z]', password))
    has_digits = bool(re.search(r'[0-9]', password))
    has_symbols = bool(re.search(r'[^A-Za-z0-9]', password))

    pool = 0
    if has_upper:
        pool += 26
    if has_lower:
        pool += 26
    if has_digits:
        pool += 10
    if has_symbols:
        pool += 32  # approximate symbol count

    if pool == 0:
        return 0

    return math.log2(pow(pool, len(password)))
